File: backend/internal/api/dto/event.go
package dto

import (
	"time"

	"github.com/daggix02-m/event_nu/backend/internal/domain"
)

type CreateEventRequest struct {
	VenueID       *string `json:"venue_id"`
	CategoryID    *string `json:"category_id"`
	Title         string  `json:"title"`
	Description   string  `json:"description"`
	StartsAt      string  `json:"starts_at"` // RFC3339
	EndsAt        *string `json:"ends_at"`   // RFC3339, optional
	PriceIsFree   bool    `json:"price_is_free"`
	PriceDisplay  string  `json:"price_display"`
	ActionType    string  `json:"action_type"`
	ActionTarget  string  `json:"action_target"`
	MaxAttendees  *int    `json:"max_attendees"`
	PosterMediaID *string `json:"poster_media_id"`
	TeaserMediaID *string `json:"teaser_media_id"`
}

type EventDTO struct {
	ID               string     `json:"id"`
	OrganizerID      string     `json:"organizer_id"`
	VenueID          *string    `json:"venue_id"`
	CategoryID       *string    `json:"category_id"`
	Title            string     `json:"title"`
	Description      string     `json:"description"`
	StartsAt         time.Time  `json:"starts_at"`
	EndsAt           *time.Time `json:"ends_at"`
	PriceIsFree      bool       `json:"price_is_free"`
	PriceDisplay     string     `json:"price_display"`
	ActionType       string     `json:"action_type"`
	Status           string     `json:"status"`
	ModerationStatus string     `json:"moderation_status"`
	MaxAttendees     *int       `json:"max_attendees"`
	LikeCount        int        `json:"like_count"`
	LikedByMe        bool       `json:"liked_by_me"`
	SavedByMe        bool       `json:"saved_by_me"`
	SavedFolderID    *string    `json:"saved_folder_id"`
	PosterURL        string     `json:"poster_url"`
	TeaserURL        string     `json:"teaser_url"`
	CreatedAt        time.Time  `json:"created_at"`
}

// AdminEventDTO is the moderation queue row: the base event fields without the
// per-user like/save state (meaningless for an admin list) and without the
// media URL joins (fetched on demand when a poster is needed).
type AdminEventDTO struct {
	ID               string     `json:"id"`
	OrganizerID      string     `json:"organizer_id"`
	VenueID          *string    `json:"venue_id"`
	CategoryID       *string    `json:"category_id"`
	Title            string     `json:"title"`
	Description      string     `json:"description"`
	StartsAt         time.Time  `json:"starts_at"`
	EndsAt           *time.Time `json:"ends_at"`
	PriceIsFree      bool       `json:"price_is_free"`
	PriceDisplay     string     `json:"price_display"`
	ActionType       string     `json:"action_type"`
	Status           string     `json:"status"`
	ModerationStatus string     `json:"moderation_status"`
	MaxAttendees     *int       `json:"max_attendees"`
	CreatedAt        time.Time  `json:"created_at"`
}

func NewAdminEventDTO(e *domain.Event) AdminEventDTO {
	return AdminEventDTO{
		ID:               e.ID,
		OrganizerID:      e.OrganizerID,
		VenueID:          e.VenueID,
		CategoryID:       e.CategoryID,
		Title:            e.Title,
		Description:      e.Description,
		StartsAt:         e.StartsAt,
		EndsAt:           e.EndsAt,
		PriceIsFree:      e.PriceIsFree,
		PriceDisplay:     e.PriceDisplay,
		ActionType:       e.ActionType,
		Status:           e.Status,
		ModerationStatus: e.ModerationStatus,
		MaxAttendees:     e.MaxAttendees,
		CreatedAt:        e.CreatedAt,
	}
}

File: flutter/lib/features/discovery/data/event.dart
import 'package:intl/intl.dart';

import '../../../core/ext/json_utils.dart';

class Event {
  const Event({
    required this.id,
    required this.organizerId,
    this.venueId,
    this.categoryId,
    required this.title,
    required this.description,
    required this.startsAt,
    this.endsAt,
    required this.priceIsFree,
    required this.priceDisplay,
    required this.actionType,
    required this.status,
    required this.moderationStatus,
    this.maxAttendees,
    required this.likeCount,
    required this.likedByMe,
    required this.savedByMe,
    this.savedFolderId,
    this.posterUrl,
    this.teaserUrl,
    required this.createdAt,
  });

  final String id;
  final String organizerId;
  final String? venueId;
  final String? categoryId;
  final String title;
  final String description;
  final DateTime startsAt;
  final DateTime? endsAt;
  final bool priceIsFree;
  final String priceDisplay;
  final String actionType;
  final String status;
  final String moderationStatus;
  final int? maxAttendees;
  final int likeCount;
  final bool likedByMe;
  final bool savedByMe;
  final String? savedFolderId;
  final String? posterUrl;
  final String? teaserUrl;
  final DateTime createdAt;

  bool get isPublished => status == 'published' && moderationStatus == 'approved';

  /// Human label like "Today, 6:30 PM" / "Sat, Sep 20" for the feed.
  String get whenLabel {
    final now = DateTime.now();
    final day = DateTime(startsAt.year, startsAt.month, startsAt.day);
    final today = DateTime(now.year, now.month, now.day);
    final diff = day.difference(today).inDays;
    final time = DateFormat.jm().format(startsAt.toLocal());
    final date = switch (diff) {
      0 => 'Today',
      1 => 'Tomorrow',
      _ => DateFormat('EEE, MMM d').format(startsAt.toLocal()),
    };
    return '$date · $time';
  }

  /// Price pill: e.g. "Free" or "300 ETB" from the server's display string.
  String get priceLabel => priceIsFree ? 'Free' : (priceDisplay.isEmpty ? 'Paid' : priceDisplay);

  factory Event.fromJson(Map<String, dynamic> json) {
    return Event(
      id: json.strOr('id', ''),
      organizerId: json.strOr('organizer_id', ''),
      venueId: json.str('venue_id'),
      categoryId: json.str('category_id'),
      title: json.strOr('title', ''),
      description: json.strOr('description', ''),
      startsAt: json.date('starts_at') ?? DateTime.fromMillisecondsSinceEpoch(0),
      endsAt: json.date('ends_at'),
      priceIsFree: json.boolOr('price_is_free', json.boolOr('is_free', true)),
      priceDisplay: json.strOr('price_display', ''),
      actionType: json.strOr('action_type', ''),
      status: json.strOr('status', ''),
      moderationStatus: json.strOr('moderation_status', ''),
      maxAttendees: json.integer('max_attendees'),
      likeCount: json.intOr('like_count', 0),
      likedByMe: json.boolOr('liked_by_me', false),
      savedByMe: json.boolOr('saved_by_me', false),
      savedFolderId: json.str('saved_folder_id'),
      posterUrl: _imageUrl(json.str('poster_url')),
      teaserUrl: _imageUrl(json.str('teaser_url')),
      createdAt: json.date('created_at') ?? DateTime.fromMillisecondsSinceEpoch(0),
    );
  }

  static String? _imageUrl(String? url) {
    if (url == null || url.isEmpty) return null;
    if (url.startsWith('http')) return url;
    return null;
  }
}